e2e/whisper/whisper_mailbox_test.go

```diff
@@ -86,13 +86,18 @@ func (s *WhisperMailboxSuite) TestRequestMessageFromMailboxAsync() {
 	//Post message
 	s.postMessageToPrivate(rpcClient, pubkey.String(), topic.String(), hexutil.Encode([]byte("Hello world!")))
 
-	//There are no messages, because it's a sender filter
+	//wait to receive message and retrieve it
```

Comment:
the code is correct obviously, but i am not sure that it addresses the root cause of this failure

for some reason it was expected that message will not be visible to a sender. but it might be that a duplicate envelope was received by a sender and the initial assumption is wrong. could you investigate how it actually ends up in this filter? and add an appropriate comment to this test,  also if i am right than L95-96 can be removed

Reply:
hey @dshulyak thanks for your review :)

> could you investigate how it actually ends up in this filter?

Sure, on it. 

I got confused by the calls to `s.getMessagesByMessageFilterID` in the flaky test, on this line https://github.com/status-im/status-go/blob/develop/e2e/whisper/whisper_mailbox_test.go#L90 it was expected to not retrieve any messages, but on this line https://github.com/status-im/status-go/blob/develop/e2e/whisper/whisper_mailbox_test.go#L117, after requesting expired messages, the same call to `s.getMessagesByMessageFilterID`, with the same filter, is expected to retrieve one message, is this correct? If so this seems to be a good way to verify that the message only came from the MailServer?

Thanks!

Reply:
Looks like the filter is correctly catching the message, both the private key and the topic match, so this condition is satisfied https://github.com/ethereum/go-ethereum/blob/master/whisper/whisperv5/filter.go#L184 Please, correct me if I'm wrong, absolute noob in terms of whisper here :)

According to the code and name of the test the goal is checking async message retrieval from a mailbox, I think that this is achieved now. IMO that another option for making sure that the message comes from the inbox could be to stop the node, send the message, and then start the node (trying to simulate an offline user), not sure if this would work though, I haven't been able to test it successfully.

I've added more comments to make clear the approach @dshulyak let me know if L95-96 should be removed, AIUI they are a bit of double checking but maybe worth keeping them.

Reply:
Ok, done.

Reply:
> Envelope that be passed to a filter needs to be added here https://github.com/ethereum/go-ethereum/blob/master/whisper/whisperv5/whisper.go#L628 . And it is called right after https://github.com/ethereum/go-ethereum/blob/master/whisper/whisperv5/api.go#L313 .

Thanks a lot for the pointers =) the envelope is indeed added at that point. 

Trying to understand how things work I've been tracing a little further and, in the `NotifyWatchers` function that determines if received messages should be added to a filter, when there's a single message (as in the test) this condition is never met https://github.com/ethereum/go-ethereum/blob/master/whisper/whisperv5/filter.go#L111 and the `MatchMessage` method that i referred to above is never called (sorry for not checking this better before). Instead, `MatchEnvelope` is called here https://github.com/ethereum/go-ethereum/blob/master/whisper/whisperv5/filter.go#L114, but in that case only the topic is checked https://github.com/ethereum/go-ethereum/blob/master/whisper/whisperv5/filter.go#L197. If that check succeeds later the key is taken into account when the envelope tries to open the message here https://github.com/ethereum/go-ethereum/blob/master/whisper/whisperv5/filter.go#L116, if this fails the message won't be added to the filter. 

So, according to all this currently only the key and the topic determine if the message should be added to a filter, please correct me if I'm wrong. Where is the sender filtering expected to be happening?

Reply:
I am not the author of this test, but maybe the initial idea was to use asym keys and then envelope will fail to open. with sym keys filter works as designed